### ISSUE TITLE
Bug: UART_RX empty check causes data dequeue

### ISSUE THREAD

**Issue by hongyihuang** [ucb-bar/Baremetal-IDE#1]
Where & what: In [bearly branch](https://github.com/ucb-bar/Baremetal-IDE/blob/bearly/workspace/bsp/bearlyml/src/bearlyml_hal_uart.c
), UART RX driver has a bug that causes data loss.

Why: When determining if the queue is empty, the receive driver actually reads in the whole 32 bit register also containing the data. It is causing the RX FIFO to be dequeued while empty is polled, losing one byte of data. 

```
while (READ_BITS(UARTx->RXDATA, UART_RXDATA_EMPTY_MSK)) {
  // return TIMEOUT;
}
*data = UARTx->RXDATA;
...
```

Action: A correct implementation should read in the 32 bit MMIO Reg at once to a variable, check if empty bit is low, then place the RX byte into the array without reading the register again.

**Comment by T-K-233** [comment]
Fixed in PR #3

**Comment by KING-VA** [comment]
Verified on BearlyML 22 that UART receive works as intended. I think we can close this issue.
